### PR TITLE
Plugins: Handle app plugin proxy routes per request

### LINKED ISSUE #47530: Plugins: Installing/uninstalling app during runtime doesn't register/unregister routes
**What happened**:
Installed an app plugin, e.g. Cloudflare Grafana App, from the UI (plugin catalog) and couldn't authenticate/use it until restarting Grafana.

**What you expected to happen**:
Could use the app directly after installation without any errors (404) etc.

**How to reproduce it (as minimally and precisely as possible)**:
1. Install an app plugin that has routes defined, e.g. Cloudflare Grafana App
2. Configure and enable the app and try to use it you'll notice you get 404's for the app's proxied requests.

**Environment**:
- Grafana version: main and all stable versions supporting installing plugins during runtime

### PR DESCRIPTION
**Pull request by marefr**
**What this PR does / why we need it**:
Rather than register each app plugin route as a separate route handles app plugin proxy routes dynamically in the same way as the datasource proxy is implemented.

To support app routes with wildcards, e.g. `/something/*` we use the web package to match routes dynamically and it seems to work as expected.

**Which issue(s) this PR fixes**:
Fixes #47530

**Special notes for your reviewer**:
Tested with a couple different app plugins and seems to work as expected. Has especially tested with the cloudflare app which uses wildcard routes.

### CODE REVIEW

File: pkg/api/pluginproxy/pluginproxy.go [hunk 1]
```diff
@@ -5,121 +5,184 @@ import (
...
+		// method match
+		if route.Method != "" && route.Method != "*" && route.Method != proxy.ctx.Req.Method {
+			continue
 		}
 
-		secureJsonData, err := secretsService.DecryptJsonData(ctx.Req.Context(), ps.SecureJSONData)
-		if err != nil {
-			ctx.JsonApiErr(500, "Failed to decrypt plugin settings", err)
-			return
+		if !web.MatchTest(route.Path, proxy.proxyPath) {
```
Comment (marefr) [inline pkg/api/pluginproxy/pluginproxy.go:58]: Think I have to use Tree.Match to get the matched `params` and pick out the `*` param and set the proxy path to that one.
https://github.com/grafana/grafana/blob/db9c9b5354e0e58fbc733fe11a6f229ab6be6dd8/pkg/web/tree.go#L377

File: pkg/api/api.go [hunk 2]
```diff
@@ -339,6 +339,8 @@ func (hs *HTTPServer) registerRoutes() {
 		apiRoute.Any("/plugins/:pluginId/resources", authorize(reqSignedIn, ac.EvalPermission(plugins.ActionAppAccess, pluginIDScope)), hs.CallResource)
 		apiRoute.Any("/plugins/:pluginId/resources/*", authorize(reqSignedIn, ac.EvalPermission(plugins.ActionAppAccess, pluginIDScope)), hs.CallResource)
 		apiRoute.Get("/plugins/errors", routing.Wrap(hs.GetPluginErrorsList))
+		apiRoute.Any("/plugin-proxy/:pluginId/*", authorize(reqSignedIn, ac.EvalPermission(plugins.ActionAppAccess, pluginIDScope)), hs.ProxyPluginRequest)
+		apiRoute.Any("/plugin-proxy/:pluginId", authorize(reqSignedIn, ac.EvalPermission(plugins.ActionAppAccess, pluginIDScope)), hs.ProxyPluginRequest)
```
Comment (wbrowne) [inline pkg/api/api.go:343]: Just confirming we support requests at the root - do we have examples of this where route.Path is set to `""` or `"/"`?

Comment (marefr) [reply pkg/api/api.go:343]: Haven't found any, but in theory we could support it. Same as datasource proxy. Do you see any issue with that?

Comment (wbrowne) [reply pkg/api/api.go:343]: I don't think so... Have just checked against GCOM in any case for curiosity and don't think it exists. Can keep or omit I reckon

File: pkg/api/api.go [hunk 3]
```diff
@@ -338,6 +338,8 @@ func (hs *HTTPServer) registerRoutes() {
...
```
[the remark this answers is not in the document]
Comment (gamab) [reply pkg/api/api.go:343]: Yes it does :+1:

**Comment by tolzhabayev** [comment]
Just fyi - I removed the 9.1 milestone, feel free to add a new one once it's clear when this PR will be merged

File: pkg/api/api.go (hunk 2 above)
Comment (marefr) [reply pkg/api/api.go:369]: I'll suggest we stick with this just in case

**Review by wbrowne — APPROVED** [review]

**Comment by bergquist** [comment]
Sweet! This solves https://github.com/grafana/grafana/pull/53930 as well :)

**Comment by marefr** [comment]
Nice :+1:

**Pull request merged**